Modules/src/StatusAggregator.cc

```diff
@@ -1,114 +1,204 @@
+            break;
+          }
+        }
+        if(skip) continue;
 
-        statusInput.emplace_back(this, moduleName, "", "");
+        // aggregate the StatusAggregator's result
+        auto node = VariableNetworkNode(aggregator->_output.status);
+        _inputs.emplace_back(this, node.getName(), "", std::unordered_set<std::string>{tagInternalVars});
+        node >> _inputs.back();
+        return;
       }
-      else if(dynamic_cast<ModuleGroup*>(module)) {
-        std::cout << "Found ModuleGroup " << module->getName() << std::endl;
-        // Process level below
-        scanAndPopulateFromHierarchyLevel(module->getAccessorList());
+    }
+
+    // If no (matching) StatusAggregator is found, recurse into sub-modules
+    for(const auto& submodule : ml) {
+      if(dynamic_cast<StatusAggregator*>(submodule) != nullptr) continue; // StatusAggregators are already handled
+      scanAndPopulateFromHierarchyLevel(*submodule, namePrefix + "/" + submodule->getName());
+    }
+  }
+
+  /********************************************************************************************************************/
+
+  int StatusAggregator::getPriority(StatusOutput::Status status) {
+    auto is = int32_t(status);
+
+    // Note, without alteration of the value, we have the PriorityMode::fwko
+    if(_mode == PriorityMode::fwok) {
+      // swap 0 <-> 1
+      if(is == 0) {
+        is = 1;
       }
-      else {
-        //FIXME (wip) Some warning if a module could not be properly classified
-        std::cout << "StatusAggregator WARNING: Module" << module->getName() << " was not detected." << std::endl;
+      else if(is == 1) {
+        is = 0;
       }
     }
+    else if(_mode == PriorityMode::ofwk) {
+      // subtract 1, then change -1 into 3
+      --is;
+      if(is == -1) is = 3;
+    }
+    else if(_mode == PriorityMode::fw_warn_mixed) {
+      // change 0,1 into -1 (cf. special meaning of -1)
+      if(is <= 1) is = -1;
+    }
 
-    // 2. TODO Add status inputs
+    return is;
   }
+
+  /********************************************************************************************************************/
+
+  void StatusAggregator::mainLoop() {
+    auto rag = readAnyGroup();
+    bool initialValue = true;
+    while(true) {
+      // find highest priority status of all inputs
+      StatusOutput::Status status;
+      bool statusSet = false; // flag whether status has been set from an input already
+      for(auto& input : _inputs) {
+        auto prio = getPriority(input);
+        if(!statusSet || prio > getPriority(status)) {
+          status = input;
+          statusSet = true;
+        }
+        else if(prio == -1) { //  -1 means, we need to warn about mixed values
+          if(getPriority(status) == -1 && input != status) {
+            status = StatusOutput::Status::WARNING;
+          }
+        }
+      }
+      assert(statusSet);
+
+      // write status only if changed, but always write initial value out
```

Comment:
This is not true. When the status changes, it will inherit the correct version number and hence it can be correlated. If it does not change, what do you want to correlate?

Reply:
Well, modules usually write their outputs unconditionally of its content. One could argue that the monitor should always update to indicate that it has updated the status (changing version number indicating that module is still working.

But I agree that the status is a "slowly changing" variable. You also would not send it with 10 Hz via ZMQ but only when it changes. So we do the same in the application.